zip-0312.rst

```diff
@@ -1,7 +1,451 @@
+[#frost-protocol]_. The Coordinator is responsible for sending the message to be
+signed to all participants, and to aggregate the signature shares.
+
+With those considerations in mind, the threat model considered in this ZIP is:
+
+- The Coordinator is trusted with the privacy of the transaction (which includes
+  the unlinkability property). A rogue Coordinator will be able to break
+  unlinkability and privacy, but should not be able to create signed transactions
+  without the approval of ``MIN_PARTICIPANTS`` participants, as specified in FROST.
+- All key share holders are also trusted with the privacy of the transaction,
+  thus a rogue key share holder will be able to break its privacy and unlinkability.
+
+
+Non-requirements
+================
+
+- This ZIP does not support removing the Coordinator role, as described in
+  [#frost-removingcoordinator]_.
+- This ZIP does not prevent key share holders from linking the signing operation to a
+  transaction in the blockchain.
+- Like the FROST specification [#FROST]_, this ZIP does not specify a key generation
+  procedure; but refer to that specification for guidelines.
+- Network privacy is not in scope for this ZIP, and must be obtained with other
+  tools if desired.
+
+
+Specification
+=============
+
+Algorithms in this section are specified using Python pseudo-code, in the same
+fashion as the FROST specification [#FROST]_.
+
+The types Scalar, Element, and G are defined in [#frost-primeordergroup]_, as
+well as the notation for elliptic-curve arithmetic, which uses the additive
+notation. Note that this notation differs from that used in the Zcash Protocol
+Specification. For example, ``G.ScalarMult(P, k)`` is used for scalar
+multiplication, where the protocol spec would use :math:`[k] P` with the group
+implied by :math:`P`.
+
+An additional per-ciphersuite hash function is used, denote ``HR(m)``, which
+receives an arbitrary-sized byte string and returns a Scalar. It is defined
+concretely in the Ciphersuites section.
+
+
+Re-randomizable FROST
+---------------------
+
+To add re-randomization to FROST, follow the specification [#FROST]_ with the
+following modifications.
+
+
+Key Generation
+''''''''''''''
+
+While key generation is out of scope for this ZIP and the FROST spec [#FROST]_,
+it needs to be consistent with FROST, see [#frost-tdkg]_ for guidance. The spend
+authorization private key :math:`\mathsf{ask}` [#protocol-spendauthsig]_ is the
+particular key that must be used in the context of this ZIP. Note that the
+:math:`\mathsf{ask}` is usually derived from the spending key
+:math:`\mathsf{sk}`, though that is not required. Not doing so allows using
+distributed key generation, since the key it generates is unpredictable. Note
+however that not deriving :math:`\mathsf{ask}` from :math:`\mathsf{sk}` prevents
+using seed phrases to recover the original secret (which may be something
+desirable in the context of FROST).
```

Comment:
```suggestion
While key generation is out of scope for this ZIP and the FROST spec [#FROST]_,
it needs to be consistent with FROST; see [#frost-tdkg]_ for general guidance.

To define a spending or viewing key that uses FROST, the Sapling and Orchard key
trees [#protocol-saplingkeycomponents]_ [#protocol-orchardkeycomponents]_ are
adjusted as follows:

- The Spend validating key :math:`\mathsf{ak}` is replaced by the FROST group
  public key `PK` [#frost-protocol]_.
- The Spend authorizing key :math:`\mathsf{ask}` is replaced by the logical
  signing key that corresponds to the group public key `PK`. By design, this
  key never exists, and instead is represented by each participant's FROST
  signing key share `sk_i`.

The remaining parts of the Sapling and Orchard key trees are generated from
a common... (TODO: Finish specifying how the other common parts of the
Sapling and Orchard key trees are derived for participants, perhaps in terms
of a common `sk` or a common HD path.)

(Old remaining content below, which might change after the above TODO.)
Note that the :math:`\mathsf{ask}` is usually derived from the spending key
:math:`\mathsf{sk}`, though that is not required. Not doing so allows using
distributed key generation, since the key it generates is unpredictable. Note
however that not deriving :math:`\mathsf{ask}` from :math:`\mathsf{sk}` prevents
using seed phrases to recover the original secret (which may be something
desirable in the context of FROST).
```

Reply:
Moved to #883.